contracts/liquidityMining/KyberRewardLocker.sol

```diff
@@ -0,0 +1,317 @@
+
+  modifier onlyRewardsContract(IERC20Ext token) {
+    require(rewardContractsPerToken[token].contains(msg.sender), 'onlyRewardcontract');
+    _;
+  }
+
+  constructor(address _admin) PermissionAdmin(_admin) {}
+
+  /**
+   * @notice Add a whitelisted rewards contract
+   */
+  function addRewardsContract(IERC20Ext token, address _rewardContract) external onlyAdmin {
+    require(
+      rewardContractsPerToken[token].length() < MAX_REWARD_CONTRACTS_SIZE,
+      'rewardContracts is too long'
+    );
+    require(rewardContractsPerToken[token].add(_rewardContract), '_rewardContract is added');
+
+    emit RewardContractAdded(_rewardContract, true);
+  }
+
+  /**
+   * @notice Remove a whitelisted rewards contract
+   */
+  function removeRewardsContract(IERC20Ext token, address _rewardContract) external onlyAdmin {
+    require(rewardContractsPerToken[token].remove(_rewardContract), '_rewardContract is removed');
+
+    emit RewardContractAdded(_rewardContract, false);
+  }
+
+  function setSlashingTarget(IERC20Ext token, address target) external onlyAdmin {
+    slashingTargets[token] = target;
+
+    emit SetSlashingTarget(token, target);
+  }
+
+  function setLockTime(IERC20Ext token, uint256 _lockTime) external onlyAdmin {
+    lockTime[token] = _lockTime;
+
+    emit SetLockTime(token, _lockTime);
+  }
+
+  function lock(
+    IERC20Ext token,
+    address account,
+    uint256 quantity
+  ) external override {
+    lockWithStartTime(token, account, quantity, _blockTimestamp());
+  }
+
+  function lockWithStartTime(
+    IERC20Ext token,
+    address account,
+    uint256 quantity,
+    uint256 startTime
+  ) public override onlyRewardsContract(token) {
+    require(quantity != 0, 'Quantity cannot be zero');
+
+    // transfer token from reward contract to lock contract
+    token.safeTransferFrom(msg.sender, address(this), quantity);
+
+    VestingSchedules storage schedules = accountsVestingSchedules[account][token];
+    uint256 schedulesLength = schedules.length;
+
+    uint256 endTime = startTime.add(lockTime[token]);
+
+    if (schedulesLength == 0) {
+      accountEscrowedBalance[account][token] = quantity;
+    } else {
+      /* Disallow adding new vested XTK earlier than the last one. */
+      require(schedules.data[schedulesLength - 1].endTime < endTime, 'not linear vesting');
+      accountEscrowedBalance[account][token] = accountEscrowedBalance[account][token].add(
+        quantity
+      );
+    }
+    // append to storage, the schedule data
+    schedules.data[schedulesLength] = VestingSchedule({
+      startTime: startTime.toUint64(),
+      endTime: endTime.toUint64(),
+      quantity: quantity.toUint128()
+    });
+    schedules.length = schedulesLength + 1;
+
+    emit VestingEntryCreated(token, account, _blockTimestamp(), quantity);
+  }
+
+  /**
+   * @notice Allow a user to withdraw any XTK in their schedule that have vested.
+   */
+  function vestAll(IERC20Ext token) external returns (uint256) {
```

Comment:
maybe can delete all vesting schedules of the user at the end, since the data is not really needed..?